### DISCUSSION TITLE
help request:  The route occasionally returns a 404 page not found

### DISCUSSION THREAD

**Discussion by wangfeng22** [apache/apisix#8288] [Q&A]
### Description

I have configured several routes, and everything going well. But when I added a new route, sometimes this route returned 404 page not found. That's really abnormal.In fact, the  back-end  didn't handle the 404 requests. 
I have restarted apisix, but this problem still exists. Only the domain name added later has this problem.

ps: The Route support http and https. The Upstream is configured like http://`domain`

### Environment

- APISIX version (run `apisix version`): 2.13.2
- Operating system (run `uname -a`):
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`):
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`):
- APISIX Dashboard version, if relevant:
- Plugin runner version, for issues related to plugin runners:
- LuaRocks version, for installation issues (run `luarocks --version`):

**Comment by juststillthinking** [comment]
Could you provide a minimal reproduction use case with detailed steps？

**Comment by tzssangglass** [comment]
maybe fix by: #7466 and release in v2.15.0

**Comment by wangfeng22** [comment]
> maybe fix by: #7466 and release in v2.15.0

I have studied this issue for many days, and I find it difficult to understand the explanation of this problem. Why can't I restart Apisix to solve this problem. @tzssangglass

**Comment by wangfeng22** [comment]
> Could you provide a minimal reproduction use case with detailed steps？

1. add several routes like routerA, routerB, routerC, and so on. The router support HTTP and HTTPS. The upstream is configured like http://`domain`.
2. add routerD. Then a strange thing happened, routerD occasionally returns `404 page not found`. In fact, the upstream of routerD don't handle the 404 requests.

I don't know what returns `404 page not found`. routerA, routerB and routerC are working well.  I have restarted apisix, but this problem still exists. @monkeyDluffy6017

**Comment by wangfeng22** [comment]
> maybe fix by: #7466 and release in v2.15.0

<img width="1424" alt="image" src="https://user-images.githubusercontent.com/42538040/200737195-0e8c486a-2aaa-46cb-be51-00e96444eff0.png">
 
The 404 response is returned right now, and it doesn't get a response from upstream. You can see it from No number.
@tzssangglass

**Comment by wangfeng22** [comment]
@tokers Could you help me see see?  Looking forward to your reply.

**Comment by tzssangglass** [comment]
> The 404 response is returned right now, and it doesn't get a response from upstream. You can see it from No number.

Are you saying that this problem still exists in 2.15.0? Or are you insisting on reproducing and working on this issue on 2.13.2?

**Comment by wangfeng22** [comment]
> 

I insist on reproducing and working on this issue on 2.13.2. I really want to know why this happened. This problem can be reproduced in the case of pressure measurement. It is difficult to understand the explanation of this problem in this issue #7466  @tzssangglass

**Comment by juststillthinking** [comment]
I checked the code, and will show you how this happened.
If without this fix https://github.com/apache/apisix/pull/7466, we assume that the first request is https, and the second request is http. we will get `pool_opt: {\"pool\":\"127.0.0.1#1983#a.com\",\"pool_size\":4}` in req1, get `pool_opt: {\"pool\":\"127.0.0.1#1983#a.com\",\"pool_size\":8}` in req2, the http req will get the https req's pool, and it's not expected. the http req will use the connection from the https connection pool, and get 404 response.

**Reply by wangfeng22** [reply]
I config no HTTPS route or HTTPS upstream. Then I restarted apisix, and the problem still exists.

**Reply by wangfeng22** [reply]
> I checked the code, and will show you how this happened. If without this fix #7466, we assume that the first request is https, and the second request is http. we will get `pool_opt: {\"pool\":\"127.0.0.1#1983#a.com\",\"pool_size\":4}` in req1, get `pool_opt: {\"pool\":\"127.0.0.1#1983#a.com\",\"pool_size\":8}` in req2, the http req will get the https req's pool, and it's not expected. the http req will use the connection from the https connection pool, and get 404 response.

@monkeyDluffy6017 That confused me a few days. I really want to know why this happened.   The route get a response from another upstream. It seems no relevant to HTTPS.

**Reply by juststillthinking** [reply]
@wangfeng22 could you provide a minimal reproduction use case like https://github.com/apache/apisix/pull/7466/files?

**Reply by juststillthinking** [reply]
or tell me how to reproduce step by step from the beginning?

**Reply by wangfeng22** [reply]
1. add route A
2. add route B
3. Stress Test Route A, and request route B at the same time, route B will receive the response of A's upstream.
@monkeyDluffy6017 This one has some difficulties to reproduce the problem, I just describe the phenomenon.

**Reply by juststillthinking** [reply]
The scenario you listed is the most basic function of apisix, if there is a problem with this scenario, there will be a lot of user feedback, did you miss any condition?
Do you have keepalive enabled? If so, can you turn off keepalive and try it out?
```
            local code, body = test('/apisix/admin/upstreams/1',
                ngx.HTTP_PUT,
                [[{
                    "scheme": "https",
                    "type": "roundrobin",
                    "nodes": {
                        "127.0.0.1:1983": 1
                    },
                    "pass_host": "rewrite",
                    "upstream_host": "a.com",
                    "keepalive_pool": {                            <=========== here
                        "size": 4
                    }
                }]]
            )
```

**Reply by wangfeng22** [reply]
> The scenario you listed is the most basic function of apisix, if there is a problem with this scenario, there will be a lot of user feedback, did you miss any condition? Do you have keepalive enabled? If so, can you turn off keepalive and try it out?
> 
> ```
>             local code, body = test('/apisix/admin/upstreams/1',
>                 ngx.HTTP_PUT,
>                 [[{
>                     "scheme": "https",
>                     "type": "roundrobin",
>                     "nodes": {
>                         "127.0.0.1:1983": 1
>                     },
>                     "pass_host": "rewrite",
>                     "upstream_host": "a.com",
>                     "keepalive_pool": {                            <=========== here
>                         "size": 4
>                     }
>                 }]]
>             )
> ```

I have configured keepalive_timeout as 0 in `config.yaml`, but keepalive_pool is configured as defalut by apisix-dashboard. 
@monkeyDluffy6017

**Reply by juststillthinking** [reply]
use `curl` instead of `apisix-dashboard`

**Reply by wangfeng22** [reply]
The following is upstream config in etcd.
```
{"hash_on":"vars","desc":"api-add","type":"roundrobin","keepalive_pool":{"size":320,"idle_timeout":60,"requests":1000},"timeout":{"read":600,"send":600,"connect":600},"create_time":1667787147,"name":"domain.com","nodes":{"domain:80":1},"id":"00000000000000001868","update_time":1667787147,"scheme":"http","pass_host":"pass"}
```

**Reply by daoleno** [reply]
I encountered the same issue, @wangfeng22  have you found a solution?
* Occasionally encountering `404 route not found` error
* Occasionally being routed to a different upstream.
